### PR TITLE
added streaming support

### PR DESCRIPTION
**Pull request by marconi**

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -40,7 +44,25 @@ const useAbortableFetch = <T>(
...
+        (async () => {
+          while (true) {
+            const { value, done } = await reader.read();
+            if (done) {
+              break;
+            }
+
+            setState((oldState: FetchState) => ({
+              ...oldState,
+              data: value,
```
Comment (mauricedb) [inline src/index.ts]: Should this create a new Uint8Array with the already received chunks and the new one appended?

File: src/index.ts [hunk 2]
```diff
@@ -7,25 +7,29 @@ import {
   SetStateAction
 } from 'react';
 
+interface RequestInitStreaming extends RequestInit {
```
[the remark this answers is not in the document]
Comment (marconi) [reply src/index.ts]: Right, I think this is actually a better way so we don't need the `stream` option. Will play around with it to see what other indicators can be used to infer if streaming or not. And yes for json, we can just check the content-type.

File: src/index.ts (hunk 1 above)
Comment (marconi) [reply src/index.ts]: I tried that first but it actually kept on re-instantiating the state so I only got the latest log. For my use-case, I handled the aggregation from the caller of the hooks instead.

File: src/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (mauricedb) [reply src/index.ts]: I was searching for some but can't really find them. I would say that structured formats with a clear start and end like JSON, XML should not be streamed, even with `Transfer-Encoding: chunked` as a partial file is invalid. Formats like text, YAML, binary could be streamed. But I guess even then the client has to know when the last chunk has arrived and the response is complete.

File: src/index.ts (hunk 1 above)
Comment (mauricedb) [reply src/index.ts]: The `oldState` contains the `data` loaded so far.

**Comment by marconi** [comment]
> Do you know of a public API endpoint I can use to test this?

Here's a Go server that just infinitely streams chunks, https://gist.github.com/marconi/fbd85020678a0cd9cf973d7bf79e30c6.

File: src/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (marconi) [reply src/index.ts]: Agreed, streaming is dictated by the server. But even then, there's actually no reliable way to detect it atm as `Transfer-Encoding` header isn't even available on cors request. See https://developers.google.com/web/updates/2015/03/introduction-to-fetch#response_types. 

If somehow you have access to the server, you can expose the `Transfer-Encoding` header via https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#Access-Control-Expose-Headers but that's still unreliable since you're talking to different service, and don't even have access to it to be able to expose some headers.

So I think for one-off responses like json, xml, etc. we can infer those by checking the `content-type` header then calling `.json()` or `.text()` accordingly but that's about it. For everything else we return the the body (ReadableStream) and let the caller decide. Thoughts?

File: src/index.ts (hunk 1 above)
Comment (marconi) [reply src/index.ts]: 👍 got it working now

File: src/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (mauricedb) [reply src/index.ts]: It seems strange if a server allows for CORS calls, set the `Transfer-Encoding: chunked` header but then doesn't set the `Access-Control-Expose-Headers` for the client to use. But then strange things happen 😞 	

I would suggest using streaming makes sense if the client can read the `Transfer-Encoding: chunked` header and the `Content-Type` makes sense for streaming. If not use the `.json()` or `.text()` depending on the `Content-Type`.

File: src/index.ts (hunk 1 above)
Comment (mauricedb) [reply src/index.ts]: 💯 	👍

File: src/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (marconi) [reply src/index.ts]: So just to be clear, we can't use both `Transfer-Encoding` and `Content-Type` to decide whether to stream or not. First, `Transfer-Encoding` isn't exposed to us and second, `Content-Type` has nothing to do with streaming or not since it could be `text/plain` but still streaming so we can't rely on that either.

But we could do:
1. pass option `{ streaming: true }` - return aggregated chunks as ArrayBuffer
2. no option - return `rsp.body` promise and let caller decide to call `.json()` or `.text()` or however they want to handle the data

**Comment by marconi** [comment]
Updated so it now returns the following:

1. with `{streaming: true}` -> returns `ArrayBuffer` of concatenated chunks
2. without `{streaming: true}` -> returns `Response` so caller can decide to either call `.json()` or `.text()` or even go the reader `.body` if they need to.

**Comment by mauricedb** [comment]
Sorry for the delay, a really busy week for me.

Not really happy with this approach. By exposing the response object the developer using this needs to call `response.json()` in most cases and as that return a promise that is not a simple call you can do when rendering a React component. It also means not being able to return a typed response anymore, another thing I don't like.

I have not found a public facing server that uses chucked encoding and serves other data formats that JSON. With JSON data you need to wait for the complete result in order to parse it. So the use cases seem to be very limited.

With a simple test server using chunking and a `text/plain` mime format works well enough and if I add the `Access-Control-Expose-Headers: *` header the fetch client can read the `Transfer-Encoding: chunked` value just fine and decide on the appropriate action and pass on the individual chunks.

So to summarize I am not going to merge this PR. 

While I like the idea of supporting chunked transfers the is a very limited use case and I am not willing to sacrifice the simple API for the main use case. I will accept a PR to add chunking if the main API stays simple and it is only exposes the `Int8Array` and multiple updates is the server indicates that it is chunking using `Transfer-Encoding: chunked` and the `Content-Type` makes sense to chunk.

**Comment by marconi** [comment]
Totally understand and no worries, atm its just not always possible to have access to the server that you can set the `Access-Control-Expose-Headers: *` which means you can't access `Transfer-Encoding: chunked`. So for now I added this https://www.npmjs.com/package/use-abortable-stream-fetch strictly for streaming purpose.

**Comment by mauricedb** [comment]
👍
